feat: Add Lido Finance as a Vault

polarvoid opened the pull request:
I worked on the assigned task on including Lido Finance as a Navigator, and have completed an initial implemention that I would like to pass on to review.

A summary of my changes are as follows:

- Added a folder called ```lido``` to the ```src``` directory.
- Added lido implementation inside said folder.
- Added a test file in the ```test``` folder, with 4 tests.
- Ensured all 4/4 tests pass.
- Added an export to ```src/index.ts``` for lido.
- Modified ```package.json``` to include test for lido.

Things that need to be worked on further:
- VaultInfo and DepositorInfo are currently shells around IVaultInfo and IDepositorInfo, respectively.
- There is an error I faced with a custom layout: [Solana Program Library Issue #2683](https://github.com/solana-labs/solana-program-library/issues/2683), which would require us to update the version of the Solana web3 library being used.
- An issue with the current implementation of Lido being possibly incompatible with the other DeFi protocols due to a mismatch in numeric type (bigint vs BN).
- Review of the ```LIDO_LAYOUT```, which is not currently being used, but could be useful to get/store data regarding stSOL.
- Addresses used while testing are actual addresses picked randomly. I am unsure of the best practice here.

Dappio-emerson commented on src/lido/infos.ts:
```diff
@@ -1,28 +1,103 @@
+    return vault;
+  }
+
+  static parseVault(data: Buffer, vaultId: PublicKey): types.VaultInfo {
+    // Decode the Token data using AccountLayout
+    const decodeData = AccountLayout.decode(data);
+    const { mint, amount } = decodeData;
```
I think you might misunderstand what `vault` and `depositor` stand for.

`Vault` should store vault relevant data which `LIDO_LAYOUT` is the info need to stored here.
After user deposit their SOL to vault, Lido will mint stSOL to user instead of creating some user specific account, so `Depositor` will be their token account.

polarvoid replied:
The talk with you was super helpful, @Dappio-emerson 

Thanks a lot :)

polarvoid commented:
I've made the changes to `LIDO_LAYOUT` and updated the infos to work as intended now.

I've also added two new test cases:
- test for the maintainers and log their Pubkeys
- test for the validators and log their Pubkeys

Dappio-emerson commented on src/lido/layout.ts:
```diff
@@ -0,0 +1,79 @@
+  [
+    u64("feeCredit"),
+    publicKey("feeAddress"),
+    seedRange("stakeSeeds"),
+    seedRange("unstakeSeeds"),
```
Since `seedRange` will only be used right here, it's a bit redundant imo.
Replace it with below is fine.

```
struct([u64("begin"), u64("end")], "stakeSeeds");
struct([u64("begin"), u64("end")], "unstakeSeeds");
```

Dappio-emerson commented on test/lido.ts:
```diff
@@ -0,0 +1,55 @@
+    const vault = await lido.infos.getVault!(connection, lido.LIDO_ADDRESS);
+    console.log(`- VaultId: ${vault.vaultId.toBase58()}`);
+    console.log(`- Vault shareMint: ${vault.shareMint.toBase58()}`);
+    // console.log(`- Vault base PDA: ${vault.base.pda.toBase58()}`);
```
Remove unused comment

Dappio-emerson replied:
Remove unused comment

Dappio-emerson commented on src/lido/layout.ts:
```diff
@@ -0,0 +1,79 @@
+  [
+    u64("feeCredit"),
+    publicKey("feeAddress"),
+    seedRange("stakeSeeds"),
+    seedRange("unstakeSeeds"),
```
Since `seedRange` will only be used right here, it's a bit redundant imo.
Replace it with below is fine.

```
struct([u64("begin"), u64("end")], "stakeSeeds");
struct([u64("begin"), u64("end")], "unstakeSeeds");
```

polarvoid replied:
Yeah, I agree with you on this. I think I got a little too excited with separating everything XD.

polarvoid replied:
Anyways, I've updated it now :)

Dappio-emerson commented on src/lido/infos.ts:
```diff
@@ -0,0 +1,126 @@
+    return this.parseDepositor(depositorAccountInfo.data, depositorId);
+  }
+
+  static getDepositorId(vaultId: PublicKey, _userKey: PublicKey, _programId: PublicKey = LIDO_PROGRAM_ID): PublicKey {
+    return vaultId;
```
1. `getDepositorId` is a function to get user's stSOL association token account(ATA) in this case. And this is how we create ATA
```
return PublicKey.findProgramAddressSync(
    [userKey.toBuffer(), TOKEN_PROGRAM_ID.toBuffer(), mint.toBuffer()],
    ASSOCIATED_TOKEN_PROGRAM_ID
  )[0];
```

polarvoid replied:
Yep, updated this

Dappio-emerson commented on src/lido/layout.ts:
```diff
@@ -0,0 +1,79 @@
+  struct([vec(MAINTAINERS_ITEM_LAYOUT, "entries"), u32("maximumEntries")], "maintainers"),
+]);
+
+export const LIDO_TOKEN_LAYOUT = struct([
+  publicKey("mint"),
+  publicKey("owner"),
+  u64("amount"),
+  u8("delegateOption"),
+  publicKey("delegate"),
+  u8("state"),
+  u8("isNativeOption"),
+  u64("isNative"),
+  u64("delegatedAmount"),
+  u8("closeAuthorityOption"),
+  publicKey("closeAuthority"),
+]);
```
Use Can remove this layout, use `AccountLayout` in spl-token is totally fine.

polarvoid replied:
Hey @Dappio-emerson , I used a custom layout here because `AccountLayout` used `bigint` and the rest of our code tends to use `BN`.

Dappio-emerson replied:
I see, thanks for your effort to keep all types consistent. However, If there's something reusable in spl-token we use it directly. On the other hand, `LIDO_TOKEN_LAYOUT` would make me feel like it's different from normal token account structure but actually it's identical. Make sense to you?

polarvoid replied:
Yeah, that makes sense :)
However what would you like to do about the type mismatch, and should I just store the depositor as an `AccountLayout`?

polarvoid replied:
Oooh alright, I've made the changes. For the `DepositorInfo` however, we'd have to use a `RawAccount` rather than a `AccountLayout`, since `AccountLayout.decode()` returns a `RawAccount`.

So what I've got going on here is this:
```ts
export interface DepositorInfo extends IDepositorInfo {
  rawAccount: RawAccount;
}
```

Dappio-emerson commented on src/lido/infos.ts:
```diff
@@ -0,0 +1,126 @@
+
+export { infos };
+
+export class VaultInfoWrapper implements IVaultInfoWrapper {
```
Please implement `getHeaviestValidator` in VaultInfoWrapper, can checkout the logic inside their SDK [here](https://github.com/lidofinance/solido-sdk/blob/6fadb02977302f73d17bfdaa74ca1479066cd584/src/unstake/getWithdrawInstruction.ts#L17)

Dappio-emerson commented on src/lido/infos.ts:
```diff
@@ -0,0 +1,126 @@
+export class VaultInfoWrapper implements IVaultInfoWrapper {
+  constructor(public vaultInfo: types.VaultInfo) {}
+
+  getApr() {
+    // TODO
+    return 0;
+  }
```
Try to fetch the ARY in `getAllVaults` and `getVault` from their API, and add an additional attribute `apr` in `VaultInfo`. 
Return `5.74`(default vault set in Lido SDK) if `this.vualtInfo.apr` is `undefined` in `getApr()`.
Can check more details [here](https://github.com/lidofinance/solido-sdk/blob/main/src/api/stakeApy.ts#L11)

polarvoid replied:
Hey, since this function will be in the `VaultInfoWrapper`, could it be better to have all the logic there?

I mean fetch it inside `VaultInfoWrapper` using the API, which I think might be better because the `apr` field could be `undefined` in `VaultInfo`, so I'm guessing the primary method of interacting with it will be through the wrapper.

polarvoid replied:
Also, could you help me out with the differences in APR and APY here? I'm a little confused between them.

Dappio-emerson replied:
> Hey, since this function will be in the VaultInfoWrapper, could it be better to have all the logic there?

Actually, we would prefer done all fetching stuff at initialization stage. But like you mention above, `apr` field might sometimes be misused and might result in `undefined`, so this idea lgtm.

> could you help me out with the differences in APR and APY here?

it would be better using `getApy` instead of `getApr` here, since the staked SOL will compound.

j-tulforov commented on src/lido/infos.ts:
```diff
@@ -0,0 +1,139 @@
+  static SOL_API_HOST = "https://sol-api-pub.lido.fi";
+
+  getApr = async () => {
+    const apr: number = await axios.get(`${VaultInfoWrapper.SOL_API_HOST}/v1/apy?since_launch`).then((res) => {
```
Hi! I highly recommend u use getApr from our [sdk](https://github.com/lidofinance/solido-sdk/blob/6fadb02977302f73d17bfdaa74ca1479066cd584/src/api/stakeApy.ts#L11)

https://docs.solana.lido.fi/frontend-integration/sdk/sdk-methods#getstakeapy

j-tulforov commented on src/lido/ids.ts:
```diff
@@ -0,0 +1,5 @@
+import { PublicKey } from "@solana/web3.js";
+
+export const LIDO_PROGRAM_ID = new PublicKey("CrX7kMhLC3cSsXJdT7JDgqrRVWGnUpX3gfEfxxU2NVLi");
+export const LIDO_ADDRESS = new PublicKey("49Yi1TKkNyYjPAFdR9LBvoHcUjuPX4Df5T5yv39w2XTn");
+export const ST_SOL_MINT_ADDRESS = new PublicKey("7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj");
```
You can import them from our [sdk](https://github.com/lidofinance/solido-sdk/blob/6fadb02977302f73d17bfdaa74ca1479066cd584/src/constants.ts#L17), why not to use it?

j-tulforov commented on src/lido/layout.ts:
```diff
@@ -0,0 +1,55 @@
+
+const MAINTAINERS_ITEM_LAYOUT = struct([publicKey("pubkey")]);
+
+export const LIDO_LAYOUT = struct([
+  u8("lidoVersion"),
+  publicKey("manager"),
+  publicKey("stSolMint"),
+  struct([u64("computedInEpoch"), u64("stSolSupply"), u64("solBalance")], "exchangeRate"),
+  u8("solReserveAuthorityBumpSeed"),
+  u8("stakeAuthorityBumpSeed"),
+  u8("mintAuthorityBumpSeed"),
+  u8("rewardsWithdrawAuthorityBumpSeed"),
+  struct(
+    [u32("treasuryFee"), u32("validationFee"), u32("developerFee"), u32("stSolAppreciation")],
+    "rewardDistribution"
+  ),
+  struct([publicKey("treasuryAccount"), publicKey("developerAccount")], "feeRecipients"),
+  METRICS_LAYOUT,
+  struct([vec(VALIDATORS_ITEM_LAYOUT, "entries"), u32("maximumEntries")], "validators"),
+  struct([vec(MAINTAINERS_ITEM_LAYOUT, "entries"), u32("maximumEntries")], "maintainers"),
+]);
```
We highly recommend you use our sdk, not to implement yourself. Because soon we will update our smart contract to v2, there will be breaking changes, but our sdk can work with both version, and know how v2 scheme looks.

ironaddicteddog commented:
Hi @j-tulforov, thank you for your kindly review on this PR. 

So why aren't Dappio `Navigator` using Lido (or any other) SDK? There are 3 main reasons:

1. `Navigator` aims to become an universal layer for instantiating DeFi protocols from scratch while keeping the core logic nice and clean.
2. `Navigator` aims to use as less dependency as possible. Since there are already 10+ protocols supported, it will be extremely messy for us to import the SDK of each protocol and handle the dependency issues.
3. `Navigator` is only a read library and aims to be as light-weight as possible. (The write functions is in another module called `Gateway`)

Hope this answers to your concern.

j-tulforov commented:
@ironaddicteddog hi! My main concern is that this code will not work after about 3 weeks, because we are migrating to second version of our protocol, and there are **breaking changes**.

In this case, our sdk will help u not to waste your time, but it ups to u. Soon we will provide migration guide.

ironaddicteddog commented:
Noted. Thanks for the headsup. I suppose we will circle back to this once it's updated. cc @Dappio-emerson @polarvoid

j-tulforov commented:
@ironaddicteddog you need to be prepared for update, after update it will not work. Right know our solution is code that works with both versions, so it will not break never.

You can see it here https://github.com/lidofinance/solido-sdk/tree/solido-v2

https://github.com/lidofinance/solido-sdk/pull/8/files

Dappio-emerson commented on test/lido.ts:
```diff
@@ -0,0 +1,59 @@
+  it(" Can get APR", async () => {
+    const vault = (await lido.infos.getVault(connection, lido.LIDO_ADDRESS)) as lido.VaultInfo;
+    const vaultWrapper = new lido.VaultInfoWrapper(vault);
+    console.log(await vaultWrapper.getApy());
```
Need to fix issue for `static` function

polarvoid replied:
Yep, I commited this by mistake. Thanks for catching it, @Dappio-emerson :)

Dappio-emerson commented on src/lido/infos.ts:
```diff
@@ -0,0 +1,153 @@
+    return VaultInfoWrapper.DEFAULT_APY;
+  }
+
+  getHeaviestValidator(): types.ValidatorInfo {
+    const validatorEntries = this.vaultInfo.validators.entries;
```
Add a test for this method.

polarvoid replied:
Done :)

polarvoid commented:
Hey y'all!

So I added support for v2.

It works by:
1. Checking Lido version:
a. Query the `LIDO_ACCOUNT` data and parse the first two bytes
b. If both bytes are set to 1, it means that `account_type` is `Lido` and `lido_version` is v2.
2. Using a different `Layout` to decode the account data, based on version.

In case `lido_version` is `v2`, we make calls to get the maintainer and validator data in `getVault()`.

This is not done in `parseVault()` because we need to make RPCs to get the data from the `ValidatorList` and `MaintainerList` accounts.

This *does mean* that the `maintainers` and `validators` fields in `VaultInfo` are **optional**. However, we check for and fetch this data in `getVault()` so it should not be a problem, *unless* a user decides to use `parseVault()` directly, i.e. without using `getVault()`.

A peculiarity is the `getVersion()` function, `lidoVersion` should be a `BN`, and according to TypeScript it is, but it is getting converted to a `number` somehow, and I do not understand why. I've added an ignore to it for now, but I would love to know why it is behaving this way.

Dappio-emerson commented on src/lido/ids.ts:
```diff
@@ -0,0 +1,5 @@
+import { PublicKey } from "@solana/web3.js";
+
+export const LIDO_PROGRAM_ID = new PublicKey("CrX7kMhLC3cSsXJdT7JDgqrRVWGnUpX3gfEfxxU2NVLi");
+export const LIDO_ADDRESS = new PublicKey("49Yi1TKkNyYjPAFdR9LBvoHcUjuPX4Df5T5yv39w2XTn");
+export const ST_SOL_MINT_ADDRESS = new PublicKey("7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj");
```
Looks like the account on `mainnet` not upgrade to v2 layout yet, please test with their `testnet` setting as below on local and make sure all test case pass.

```
export const LIDO_PROGRAM_ID = new PublicKey("6RRggYnFe2EcD543QXrE3Wxp1Kgcq8qctwRrNnvjoYsL");
export const LIDO_ADDRESS = new PublicKey("Hcqw2G2FkBhBiDZz3PXFBUKuMQiKMJmXvcfqRaXkyNXF");
export const ST_SOL_MINT_ADDRESS = new PublicKey("Bz2UPiXJmCQYT8XQHaPPXG3fywNp2TENav5dcJdX9q21");
```

Endpoint
```
const connection = new Connection("https://api.testnet.solana.com", {
    commitment: "confirmed",
    confirmTransactionInitialTimeout: 180 * 1000,
  });
```

polarvoid replied:
Yep, made a few small changes and it works now :)

Dappio-emerson commented on src/lido/infos.ts:
```diff
@@ -0,0 +1,278 @@
+  getVersion(): number {
+    // I have no idea why this gets cast to a number from a BN, and I tried explicitly casting it to a BN
+    //@ts-ignore
+    const version = (this.vaultInfo.lidoVersion as BN) + 1;
```
Try `Number(this.vaultInfo.lidoVersion)`

Dappio-emerson commented on src/lido/index.ts:
```diff
@@ -0,0 +1,90 @@
+}
+
+export interface DepositorInfo extends IDepositorInfo {
+  rawAccount: RawAccount;
```
Rename `rawAccount` to `tokenAccount` for readability.

Dappio-emerson commented on src/lido/infos.ts:
```diff
@@ -0,0 +1,278 @@
+    let vault: types.VaultInfo = this.parseVault(vaultAccountInfo.data, vaultId);
+
+    // Fetch validators if using Solido v2
+    if (vault.validatorList) {
+      const validatorListAccount = await connection.getAccountInfo(vault.validatorList);
+      if (!validatorListAccount) throw Error("Error: Cannot get validator list account");
+
+      vault.validators = this._parseValidatorListAccount(validatorListAccount.data);
+    }
+
+    // Fetch maintainers if using Solido v2
+    if (vault.maintainerList) {
+      const maintainerListAccount = await connection.getAccountInfo(vault.maintainerList);
+      if (!maintainerListAccount) throw Error("Error: Cannot get maintainer list account");
+
+      vault.maintainers = this._parseMaintainerListAccount(maintainerListAccount.data);
+    }
```
Can fetch `validatorList` and `maintainerList` together with `uitls.getMulipleAccounts()` in `@dappio-wonderland/navigator` by importing `utils`, since they might have same condition corresponding to version, so that we can reduce rpc request.

ironaddicteddog commented on src/lido/ids.ts:
```diff
@@ -0,0 +1,12 @@
+
+// Testnet addresses
+
+// export const LIDO_PROGRAM_ID = new PublicKey("6RRggYnFe2EcD543QXrE3Wxp1Kgcq8qctwRrNnvjoYsL");
```
Plz remove unused comments

ironaddicteddog commented on src/lido/infos.ts:
```diff
@@ -0,0 +1,276 @@
+    };
+  }
+
+  static _parseValidatorListAccount(data: Buffer): types.ValidatorAccountInfo {
```
Add `private`

ironaddicteddog commented on src/lido/infos.ts:
```diff
@@ -0,0 +1,276 @@
+    };
+  }
+
+  static _parseMaintainerListAccount(data: Buffer): types.MaintainerAccountInfo {
```
Add `private`

polarvoid commented:
Yep, @ironaddicteddog I've made all the changes you requested :)
